File: testdata/jennies/builders/basic_struct/TypescriptBuilder/src/basic_struct/somestruct_builder_gen.ts
import * as cog from '../cog';
import * as basic_struct from '../basic_struct';

// SomeStruct, to hold data.
export class SomeStructBuilder implements cog.Builder<basic_struct.SomeStruct> {
    private readonly internal: basic_struct.SomeStruct;

    constructor() {
        this.internal = basic_struct.defaultSomeStruct();
    }

    build(): basic_struct.SomeStruct {
        return this.internal;
    }

    // id identifies something. Weird, right?
    id(id: number): this {
        this.internal.id = id;
        return this;
    }

    uid(uid: string): this {
        this.internal.uid = uid;
        return this;
    }

    tags(tags: string[]): this {
        this.internal.tags = tags;
        return this;
    }

    // This thing could be live.
    // Or maybe not.
    liveNow(liveNow: boolean): this {
        this.internal.liveNow = liveNow;
        return this;
    }
}

File: testdata/jennies/builders/basic_struct/GoBuilder/basic_struct/somestruct_builder_gen.go
package basic_struct

import (
	cog "github.com/grafana/cog/generated/cog"
)

var _ cog.Builder[SomeStruct] = (*SomeStructBuilder)(nil)

// SomeStruct, to hold data.
type SomeStructBuilder struct {
    internal *SomeStruct
    errors map[string]cog.BuildErrors
}

func NewSomeStructBuilder() *SomeStructBuilder {
	resource := NewSomeStruct()
	builder := &SomeStructBuilder{
		internal: resource,
		errors: make(map[string]cog.BuildErrors),
	}

	return builder
}



func (builder *SomeStructBuilder) Build() (SomeStruct, error) {
	if err := builder.internal.Validate(); err != nil {
		return SomeStruct{}, err
	}

	return *builder.internal, nil
}

// id identifies something. Weird, right?
func (builder *SomeStructBuilder) Id(id int64) *SomeStructBuilder {
    builder.internal.Id = id

    return builder
}

func (builder *SomeStructBuilder) Uid(uid string) *SomeStructBuilder {
    builder.internal.Uid = uid

    return builder
}

func (builder *SomeStructBuilder) Tags(tags []string) *SomeStructBuilder {
    builder.internal.Tags = tags

    return builder
}

// This thing could be live.
// Or maybe not.
func (builder *SomeStructBuilder) LiveNow(liveNow bool) *SomeStructBuilder {
    builder.internal.LiveNow = liveNow

    return builder
}

